pkg/api/version/version.go

```diff
@@ -0,0 +1,279 @@
+	svc := &Service{
+		store:   client,
+		id:      cfg.InstanceID,
+		addr:    fmt.Sprintf("%s:%d", instanceAddr, instancePort),
+		cfg:     cfg,
+		logger:  log.With(logger, "component", "versions"),
+		cancel:  cancel,
+		ctx:     ctx,
+		version: currentQuerierVersion,
+	}
+	// The service is simple only has a running function.
+	// Stopping is manual to ensure we stop as part of the shutdown process.
+	svc.BasicService = services.NewBasicService(
+		func(_ context.Context) error { return nil },
+		svc.running,
+		func(_ error) error { return nil },
+	)
+	return svc, nil
+}
+
+func (svc *Service) running(ctx context.Context) error {
+	go svc.loop()
+	<-ctx.Done()
+	return nil
+}
+
+func (svc *Service) loop() {
+	svc.wg.Add(1)
+	ticker := time.NewTicker(heartbeatInterval)
+	defer func() {
+		ticker.Stop()
+		svc.wg.Done()
+	}()
+
+	for {
+		select {
+		case <-ticker.C:
+			if err := svc.heartbeat(svc.ctx); err != nil {
+				level.Error(svc.logger).Log("msg", "failed to heartbeat", "err", err)
+			}
+
+		case <-svc.ctx.Done():
+			level.Info(svc.logger).Log("msg", "versions is shutting down")
+			return
+		}
+	}
+}
+
+func (svc *Service) heartbeat(ctx context.Context) error {
+	return svc.store.CAS(ctx, "versions", func(in interface{}) (out interface{}, retry bool, err error) {
+		var versions *versionv1.Versions
+		if in == nil {
+			versions = newVersions().(*Versions).Versions
+		} else {
+			versions = in.(*Versions).Versions
+		}
+		current, ok := versions.Instances[svc.id]
+		if !ok {
+			current = &versionv1.InstanceVersion{}
+			versions.Instances[svc.id] = current
+		}
+		current.Addr = svc.addr
+		current.ID = svc.id
+		current.Timestamp = time.Now().UnixNano()
+		current.QuerierAPI = svc.version
+		// Now prune old instances.
+		for id, instance := range versions.Instances {
+			lastHeartbeat := time.Unix(0, instance.GetTimestamp())
+			if time.Since(lastHeartbeat) > instanceTimeout {
+				level.Warn(svc.logger).Log("msg", "auto-forgetting instance from the versions because it is unhealthy for a long time", "instance", id, "last_heartbeat", lastHeartbeat.String())
+				delete(versions.Instances, id)
+			}
+		}
+		return &Versions{
+			Versions: versions,
+		}, true, nil
+	})
+}
+
+func (svc *Service) Version(ctx context.Context, req *connect.Request[versionv1.VersionRequest]) (*connect.Response[versionv1.VersionResponse], error) {
+	value, err := svc.store.Get(ctx, "versions")
+	if err != nil {
+		return nil, err
+	}
+	versions, ok := value.(*Versions)
+	if !ok {
+		// we don't have any versions yet.
+		return connect.NewResponse(&versionv1.VersionResponse{}), nil
+	}
+	// collect the minimum querier version.
+	minQuerierVersion := uint64(math.MaxUint64)
+	for _, instance := range versions.Instances {
+		if instance.QuerierAPI < minQuerierVersion {
+			minQuerierVersion = instance.QuerierAPI
+		}
+	}
+	return connect.NewResponse(&versionv1.VersionResponse{
+		QuerierAPI: minQuerierVersion,
+	}), nil
+}
+
```

Comment:
Does this mean that instances that have not yet reported their versions are not taken into account?

Reply:
Yes and we'll report the lower version so it's safe to query

Reply:
But doesn't this mean that during the first rollout we will report higher version (1) than the lowest deployed (0)? I must be missing something. For example, we add some features that rely on the version reported by this API in v1.3/1.4/1.5. What would be the user experience during the upgrade from the current version (v1.2)?

My guess it that once _the very first_ upgraded instance reports its version (e.g 1), we will report that version because it is literally the lowest _known_ version. As a result, we will try to use new features while not all the instances are upgraded.